File: ttshare_mobile/lib/services/html_archiver.dart
import 'dart:io';
import 'package:path_provider/path_provider.dart';

class HtmlArchiver {
  /// Creates a self-contained HTML file from raw HTML
  Future<String> archive(String html, String title) async {
    final dir = await getTemporaryDirectory();
    final safeTitle = _sanitizeFileName(title);
    final fileDir = Directory('${dir.path}/$safeTitle');
    if (!await fileDir.exists()) {
      await fileDir.create(recursive: true);
    }

    final filePath = '${fileDir.path}/index.html';

    final wrappedHtml = '''
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>$title</title>
</head>
<body>
$html
</body>
</html>
''';

    await File(filePath).writeAsString(wrappedHtml);
    return filePath;
  }

  String _sanitizeFileName(String name) {
    return name
        .replaceAll(RegExp(r'[<>:"/\\|?*]'), '_')
        .replaceAll(RegExp(r'\s+'), ' ')
        .trim()
        .substring(0, name.length.clamp(1, 100));
  }
}

File: ttshare_android/app/src/main/java/com/ttshare/service/HtmlArchiver.kt
package com.ttshare.service

import android.content.Context
import java.io.File

class HtmlArchiver {

    data class ArchiveResult(
        val htmlPath: String,
        val mdPath: String?
    )

    fun archive(html: String, title: String, contentDir: File): String {
        if (!contentDir.exists()) contentDir.mkdirs()

        val htmlFile = File(contentDir, "index.html")
        val wrappedHtml = """
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>$title</title>
</head>
<body>
$html
</body>
</html>
""".trimIndent()

        htmlFile.writeText(wrappedHtml)
        return htmlFile.absolutePath
    }

    fun saveMarkdown(markdown: String, contentDir: File): String {
        if (!contentDir.exists()) contentDir.mkdirs()
        val mdFile = File(contentDir, "index.md")
        mdFile.writeText(markdown)
        return mdFile.absolutePath
    }

    fun sanitizeFileName(name: String): String {
        return name
            .replace(Regex("""[<>:"/\\|?*]"""), "_")
            .replace(Regex("\\s+"), "_")
            .take(100)
    }
}
